rfcs/0002-Flexible-Raft-with-NWR.md

```diff
@@ -0,0 +1,490 @@
+
+该项目涉及代码变更的地方可以划分为如下四个模块：
+
+- **Leader选举模块：** 一个节点想要成为leader，会经过以下几个阶段：预投票、正式投票、当选leader。所以对于preVote、electSelf、becomeLeader等等与多数派模型相关的方法都会涉及NWR模型的有关代码变更。
+- **日志复制模块：** 当leader收到客户端的事务请求或者follower与leader数据存在差距时，会调用 **Replicator#sendEntries** 去发送日志复制消息（事务消息）；而心跳消息和探测消息，则是由 **Replicator#sendEmptyEntries** 发送的。日志复制中，**NodeImpl#executeApplyingTasks 和** **NodeImpl#unsafeApplyConfiguration** 方法会涉及到多数派确认。在执行这些方法的时候，都会使用 **BallotBox#appendPendingTask** 方法来构造一个待投票的Ballot（现在叫MajorityQuorum/NWRQuorum）并放置到投票箱中。
+- **一致性读模块：** 对于一致性读模块，在raft共识算法中，读取R个节点其实体现在R个节点的心跳响应。通过R个节点的心跳，能保证这个节点一定是leader，一定拥有最新的数据，我们并不是真正需要从R个节点里面读取数据。**NodeImpl#ReadIndexHeartbeatResponseClosure** 这样的方法，可以看到执行了心跳消息的多数派确认模型的逻辑，ReadIndexHeartbeatResponseClosure构造器里面传入了quorum的值，这里我们需要对应修改为NWR模型的逻辑。
+- **成员变更模块：** 对于JRaft成员变更来讲，核心逻辑是采用单成员变更的方式，即使需要同时变更多个成员时，也是会先整理出新add与新remove的成员，再逐个进行单成员变更。其核心方法 **addPeer、removePeer、changePeers、resetPeers** 等等都会涉及NWR模型的适配。
+
+![](https://img2023.cnblogs.com/blog/2784327/202307/2784327-20230701144554871-1252032815.png)
+## Detailed Design
+
+### NodeOptions
+
+在**NodeOptions类**中，我们新增了如下三个参数：readQuorumFactor、writeQuorumFactor与enableNWRMode，分别表示读因子、写因子以及是否开启NWR模型（true），默认不开启，表示多数派确认模型（false）。
+
+```
+    /**
+     * Read Quorum's factor
+     */
+    private double                          readQuorumFactor;
+    /**
+     * Write Quorum's factor
+     */
+    private double                          writeQuorumFactor;
+    /**
+     * Enable NWRMode or Not
+     */
+    private boolean                         enableNWRMode          = false;
+```
+
+对于readQuorumFactor和writeQuorumFactor两个属性，在NodeOptions类里提供了setter和getter方法便于用户自定义配置。对于enableNWRMode属性，提供了isEnableNWRModeI()来判断是否开启NWR模型，而enableNWRMode()方法表示开启NWR模式。
+
+```
+    public double getReadQuorumFactor() {
+        return readQuorumFactor;
+    }
+    public void setReadQuorumFactor(double readQuorumFactor) {
+        this.readQuorumFactor = readQuorumFactor;
+        enableNWRMode();
+    }
+    public double getWriteQuorumFactor() {
+        return writeQuorumFactor;
+    }
+    public void setWriteQuorumFactor(double writeQuorumFactor) {
+        this.writeQuorumFactor = writeQuorumFactor;
+        enableNWRMode();
+    }
+    public boolean isEnableNWRMode() {
+        return enableNWRMode;
+    }
+    private void enableNWRMode() {
+        this.enableNWRMode = true;
+    }
+```
+
+### Node Init
+
+在NodeImpl#init时，我们首先会对NodeOptions内部的readFactor和writeFactor进行校验并且进行参数同步，如果用户只设置了readFactor和writeFactor两个参数的其中之一，那么我们需要同步这两个参数的值。
+在init方法初始化node时，会首先对NWR模式下的factor进行校验与同步。
+
+```
+if(options.isEnableNWRMode() && !checkAndResetFactor(options.getWriteQuorumFactor(),
+         options.getReadQuorumFactor())){
+     return false;
+}
+```
+
+校验与同步逻辑在方法checkAndResetFactor里：
+
+```
+    private boolean checkAndResetFactor(Double writeFactor, Double readFactor){
+        if (Objects.nonNull(readFactor) && Objects.nonNull(writeFactor)) {
+            if (readFactor + writeFactor != 1) {
```

Comment:
We can't compare integer type(`1`) with float type(`readFactor + writeFactor`) precisely. I think we can change the `readFactor` and `writeFactor` to be int types, and ensure that their sum is `10`. Their range both are `(0,10)`.

Reply:
Does this mean that if the user sets **readFactor to 4** and **writeFactor to 6**. If in a cluster with **5 nodes**, _writeQuorum_ as **5 * 0.6=3** and _readQuorum_ as **n-w+1=5-3+1=3**

Reply:
Yes， that's my point @1294566108